### PR TITLE
Validate Binary Search Tree

### PR DESCRIPTION
**Pull request by JuliaGerasymenko**
https://leetcode.com/problems/validate-binary-search-tree/
 I don't like my solution to this task. Could you suggest another variant?

### CODE REVIEW

File: validBST.js [hunk 1]
```diff
@@ -0,0 +1,11 @@
+'use strict';
+
+var isValidBST = function(root, lowerBound = -Infinity, higherBound = Infinity) {
+
+  if (!root) return true;
+
+    if (root.val >= higherBound || root.val <= lowerBound) return false;
+
+   return (!isValidBST(root.left, lowerBound, root.val) || !isValidBST(root.right, root.val, higherBound))
```
Comment (DzyubSpirit) [inline validBST.js]: To simplify try to use De Morgan law:
image

**Review by DzyubSpirit — CHANGES_REQUESTED** [review]

**Comment by DzyubSpirit** [comment]
> I don't like my solution to this task. Could you suggest another variant?

Why don't you like this solution?

P.S. It's not a full solution, right? The input is an array and not the tree ;)

**Comment by JuliaGerasymenko** [comment]
It's a full solution. 
/**
 * Definition for a binary tree node.
 * function TreeNode(val) {
 *     this.val = val;
 *     this.left = this.right = null;
 * }
 */
/**
 * @param {TreeNode} root
 * @return {boolean}
 */
The input is a tree.

File: validBST.js (hunk 1 above)
Comment (JuliaGerasymenko) [reply validBST.js]: Now, it works slower than in the previous one.

File: validBST.js [hunk 2]
```diff
@@ -1,11 +1,16 @@
...
+  higherBound = Infinity
+) {
   if (!root) return true;
 
-    if (root.val >= higherBound || root.val <= lowerBound) return false;
+  if (root.val >= higherBound || root.val <= lowerBound) return false;
 
-   return (!isValidBST(root.left, lowerBound, root.val) || !isValidBST(root.right, root.val, higherBound))
-       ? false : true;
+  return !(isValidBST(root.left, lowerBound, root.val) &&
```
Comment (DzyubSpirit) [inline validBST.js]: A ternary operator `x ? false : true` can be replaced with `!x`

**Review by DzyubSpirit — CHANGES_REQUESTED** [review]

File: validBST.js (hunk 2 above)
Comment (DzyubSpirit) [reply validBST.js]: And so `!x ? false : true` should be replaced with `x`

File: validBST.js [hunk 3]
```diff
@@ -0,0 +1,13 @@
...
+var isValidBST = function(
+  root,
+  lowerBound = -Infinity,
+  higherBound = Infinity
+) {
+  if (!root) return true;
+
+  if (root.val >= higherBound || root.val <= lowerBound) return false;
+
+  return !(isValidBST(root.left, lowerBound, root.val) && isValidBST(root.right, root.val, higherBound));
```
Comment (DzyubSpirit) [inline validBST.js]: Commented here:
https://github.com/JuliaGerasymenko/Tasks/pull/9#discussion_r352547293

**Review by DzyubSpirit — CHANGES_REQUESTED** [review]

File: validBST.js (hunk 1 above)
Comment (DzyubSpirit) [reply validBST.js]: How so?

**Review by DzyubSpirit — APPROVED** [review]

**Pull request merged**